flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/CodeGenUtils.scala

```diff
@@ -694,38 +706,67 @@ object CodeGenUtils {
     }
   }
 
+  /**
+   * Generates code for converting the given external source data type to the internal data format.
+   *
+   * Use this function for converting at the edges of the API.
+   */
   def genToInternalIfNeeded(
       ctx: CodeGeneratorContext,
-      t: DataType,
-      term: String): String = {
-    if (isInternalClass(t)) {
-      s"(${boxedTypeTermForType(fromDataTypeToLogicalType(t))}) $term"
+      sourceType: DataType,
+      externalTerm: String)
+    : GeneratedExpression = {
+    // convert external source type to internal format
+    val internalResultTerm = if (isInternalClass(sourceType)) {
+      s"(${boxedTypeTermForType(fromDataTypeToLogicalType(sourceType))}) $externalTerm"
```

Comment:
We need this cast because UDF can return `Object` and DataType with internal conversion class...
Sometimes there is some gap between the conversion class and real java function return class. It is safe to add this cast, and I think JVM can optimize this cast to not affect performance.

I think we can add comment here.

Reply:
I added a comment. But actually the gap should be filled by the converters.  I think the core problem is that `int` and `Integer` are handled the same in the converters. Even though the latter can support null und needs unboxing.

Reply:
It seems that the cast for Janino in the `BridgingSqlFunctionCallGen` also solves the cast of this method. So we can get rid of this cast here. Let's see if Travis gives green light.

Reply:
The cast is actually not necessary anymore because of the cast in `BridgingSqlFunctionCallGen`. I remove it for now.